File: Python/Searching and Sorting/06_optimum_location_for_min_distance.py
"""
Idea:
Starting from -inf to +inf, using ternary search to find the minimum distance at each iteration

( Ternary search is a divide-and-conquer search algorithm that iteratively narrows down a search space
 into "three" equal parts and eliminates one-third of the space in each iteration)
"""

import math


class Solution:
    def findOptimumCost(self, line: list[int], n: int, points: list[list[int]]) -> float:

        # calculating distance between two points...
        def distance(x, y, point):

            dis = math.sqrt((x - point[0]) ** 2 + (y - point[1]) ** 2)

            return dis

        def compute(line, points, n, x):

            summ_distance = 0
            # calculate y.. coordinate...
            # since..   ax + by +c = 0..
            # calculate y..
            if line[1] != 0:
                y = -1 * (line[0] * x + line[2]) / line[1]
            else:
                y = float("inf")
            # now, we have both x, y points ..
            # find the distances from the given points and add them

            for i in range(n):
                summ_distance += distance(x, y, points[i])

            return summ_distance

        # ternary serach...
        # finding x .. from -1e6 to 1e6 .. and find y corresponding to x from line equation

        low = -1e3  # denotes the lowest x value on x-axis
        high = 1e3  # denotes the highest x value on x_axis

        precision = 1e-6
        while (high - low) > precision:

            # both mid denotes the value of x...
            mid1 = low + (high - low) / 3
            mid2 = high - (high - low) / 3

            dis1 = compute(line, points, n, mid1)
            dis2 = compute(line, points, n, mid2)

            if dis1 < dis2:
                #  go towards lower distance point .. and eliminate other 1/3 part.
                high = mid2  # can be candidate for ans

            elif dis1 > dis2:
                low = mid1  # can be candidate for ans

            else:
                low = mid1
                high = mid2

        return compute(line, points, n, (low + high) / 2)

File: C++/Searching and Sorting/06_optimum_location_for_min_distance.cpp
/*
Idea:
Starting from -inf to +inf, using ternary search to find the minimum distance at each iteration

( Ternary search is a divide-and-conquer search algorithm that iteratively narrows down a search space
 into "three" equal parts and eliminates one-third of the space in each iteration)
*/


#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    double dist(double x, double y, pair<int, int>p) {
        return sqrt(sq(x - p.first) + sq(y - p.second));
    }
    double compute(vector<pair<int, int>>p, int n, tuple<int, int, int>l, double X) {
        double res = 0;

        // calculating Y of chosen point by line equation
        double Y = -1 * (get<2>(l) + get<0>(l) * X) / get<1>(l);
        for (int i = 0; i < n; i++) {
            res += dist(X, Y, p[i]);
        }
        return res;
    }

    // Utility method to find minimum total distance
    double findOptimumCost(tuple<int, int, int>l, vector<pair<int, int>>p, int n) {
        double low = -1e6;
        double high = 1e6;

        // loop until difference between low and high become less than EPS
        while ((high - low) > EPS) {
            // mid1 and mid2 are representative x co-ordiantes  of search space
            double mid1 = low + (high - low) / 3;
            double mid2 = high - (high - low) / 3;


            double dist1 = compute(p, n, l, mid1);
            double dist2 = compute(p, n, l, mid2);

            // if mid2 point gives more total distance, skip third part
            if (dist1 < dist2)
                high = mid2;

            // if mid1 point gives more total distance, skip first part
            else
                low = mid1;
        }

        // compute optimum distance cost by sending average of low and high as X
        return compute(p, n, l, (low + high) / 2);
    }
};

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        tuple<int, int, int>line;
        int a, b, c;
        cin >> a >> b >> c;
        line = make_tuple(a, b, c);
        int n;
        cin >> n;
        vector<pair<int, int>>points;
        for (int i = 0; i < n; i++) {
            int x, y;
            cin >> x >> y;
            points.push_back({ x, y });
        }

        Solution ob;
        double ans = ob.findOptimumCost(line, points, n);

        cout << fixed << setprecision(2) << ans << endl;
    }
    return 0;
}
